### PR TITLE
feat(developer): Range expansions

### PR DESCRIPTION
**Pull request by mcdurdin**
Relates to #2241.

Range expansions are a new language feature that reduce verbosity in Keyman Keyboard Language (.kmn) files by making it possible to collapse sequential ranges with a new `..` operator. It is envisioned that these will be used primarily in stores.

Two kinds of expansions are available: character ranges, and virtual key ranges.

The syntax is:

```
char '..' char
vkey '..' vkey
```

Where the `char` terminator is a single normal character, and the `vkey` terminator is a single virtual key. `outs` is permissible for terminators but no other statement is allowed. Whitespace is per normal Keyman keyboard language syntax.

Character ranges will replace the range with the set of characters between the terminators, based on the Unicode value of the terminators.

Examples:

```
store(alphabet) 'a' .. 'z'
store(alpha_not_bq) 'ac'..'pr'..'z'
store(capitals) U+0041 .. U+005A
```

Virtual key ranges work in a similar fashion, but with virtual keys. The starting and terminating virtual key must have the same shift state. This will be most useful for character and numeric virtual keys, as the ranges are based on the numeric value of the virtual key as defined by Windows, not its position on the keyboard.

Examples:

```
store(alphakeys) [K_A] .. [K_Z]
store(numkeys) [K_0] .. [K_9]
```

Ranges must be positive, that is, you cannot use `'z' .. 'a'`.

While this expansion is not expected to be very useful in rules, it is still permitted. `context()` and other offsets are calculated on the expanded form, not the collapsed form. Similarly, `any()`/`index()` offsets are calculated on the expanded form.

'a' .. 'z' > context(3)  c this produces 'c', not 'z'

Because ranges are expanded at compile time, ranges that are overly long will result in an error, so you cannot use, for example:

```
store(AllUnicode) U+0020 .. U+10FFFF
```

The precise maximum length of a range is dependent on the context where it is used.

The changes are entirely compiler-based; no change is made to the .kmx file format or to the Keyman apps.

### CODE REVIEW

File: windows/src/developer/kmcmpdll/Compiler.cpp [hunk 1]
```diff
@@ -2494,6 +2496,13 @@ DWORD GetXString(PFILE_KEYBOARD fk, PWSTR str, PWSTR token, PWSTR output, int ma
         err = process_set_synonym(TSS_LAYER, fk, q, tstr, &mx);
         if (err != CERR_None) return err;
         continue;
+      case 19:  // #2241
+        if (*(p + 1) != '.') return CERR_InvalidToken;
+        if (sFlag) return CERR_InvalidInVirtualKeySection;
+        p += 2;
+        err = process_expansion(fk, p, tstr, &mx, max);
```
Comment (darcywong00) [inline windows/src/developer/kmcmpdll/Compiler.cpp:2503]: I noticed the param `p` becomes `q` in l. 2675

Would this be valid?
```suggestion
        q = p+2;
        err = process_expansion(fk, q, tstr, &mx, max);
```

File: windows/src/test/unit-tests/kmcomp/test.bat [hunk 2]
```diff
@@ -39,6 +39,13 @@ call :should-fail "#4280: nul should be at start of context 2" test_4280_nul_sta
 
 call :should-pass "#4423: named code constants, various tests" test_namedcodeconstants.kmn || goto :eof
 
+call :should-pass "#2241: expansions" test_expansion.kmn || goto :eof
+call :should-pass "#2241: expansions" test_expansion_1.kmn || goto :eof
+:: the two files should be identical.
+fc /b test_expansion.kmx test_expansion_1.kmx || goto :eof
+
+call :should-fail "#2241: invalid expansions" test_expansion_invalid.kmn || goto :eof
+
```
Comment (darcywong00) [inline windows/src/test/unit-tests/kmcomp/test.bat:48]: Is the next file `test_expansion.bmp` an empty file that can be removed?

File: windows/src/test/unit-tests/kmcomp/test_expansion.kmn [hunk 3]
```diff
@@ -0,0 +1,15 @@
+c This file tests that expansions are valid, using kmdecomp to verify
+store(&NAME) 'Test Expansions'
```
Comment (darcywong00) [inline windows/src/test/unit-tests/kmcomp/test_expansion.kmn:2]: Is expansion a `store(&VERSION) "14.0"` feature?

File: windows/src/test/unit-tests/kmcomp/test_expansion_invalid.kmn [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+store(char_range_to_key) 'a' .. [K_Z]
+store(key_range_to_char) [K_A] .. 'z'
+
+store(char_range_to_statement) 'a' .. nul
+store(key_range_to_statement) [K_A] .. nul
+store(char_range_from_statement) nul .. 'a'
+store(key_range_from_statement) nul .. [K_A]
+
+store(key_range_shift_inconsistent) [SHIFT K_A] .. [K_Z]
```
Comment (darcywong00) [inline windows/src/test/unit-tests/kmcomp/test_expansion_invalid.kmn:34]: Is there a practical limit to test when the expansion buffer would overflow?

File: windows/src/developer/kmcmpdll/Compiler.cpp (hunk 1 above)
Comment (mcdurdin) [reply windows/src/developer/kmcmpdll/Compiler.cpp:2503]: No, that wouldn't work. `p` is our working variable there. We could rename the parameter in `process_expansion` though.

File: windows/src/test/unit-tests/kmcomp/test.bat (hunk 2 above)
Comment (mcdurdin) [reply windows/src/test/unit-tests/kmcomp/test.bat:48]: Yes, good catch.

File: windows/src/test/unit-tests/kmcomp/test_expansion.kmn (hunk 3 above)
Comment (mcdurdin) [reply windows/src/test/unit-tests/kmcomp/test_expansion.kmn:2]: No, per the discussion, it's a compiler only feature that is backwardly compatible with v6.0.

File: windows/src/test/unit-tests/kmcomp/test_expansion_invalid.kmn (hunk 4 above)
Comment (mcdurdin) [reply windows/src/test/unit-tests/kmcomp/test_expansion_invalid.kmn:34]: Not really. We could throw in a test for U+0020 .. U+10FFFF for fun :)

**Comment by darcywong00** [comment]
I forget the PR, but isn't there also a way to reference keys by an ISO name? (re: range expansion)

**Comment by mcdurdin** [comment]
> I forget the PR, but isn't there also a way to reference keys by an ISO name? (re: range expansion)

Yes. But it would not work very well here, because the ranges are always by virtual key value. So `[C01] .. [C03]` would not expand to `[K_A] [K_S] [K_D]` but would instead expand to `[K_A] [K_B] [K_C] [K_D]`!

**Review by darcywong00 — APPROVED** [review]
lgtm

(can delete the empty file)

File: windows/src/test/unit-tests/kmcomp/test_expansion.kmn (hunk 3 above)
Comment (jahorton) [reply windows/src/test/unit-tests/kmcomp/test_expansion.kmn:2]: ... with the v6.0 compiled format (hence, `store(&VERSION)`), but not the v6.0 compiler.

I suppose there's no flag / system store for required compiler version, huh?

Comment (mcdurdin) [reply windows/src/test/unit-tests/kmcomp/test_expansion.kmn:2]: Yeah, we don't have any flag for minimum required compiler version at present. But most programming languages don't specify that, from what I can tell, so we are probably not doing anything too awful!

**Comment by jahorton** [comment]
> > I forget the PR, but isn't there also a way to reference keys by an ISO name? (re: range expansion)
> 
> Yes. But it would not work very well here, because the ranges are always by virtual key value. So `[C01] .. [C03]` would not expand to `[K_A] [K_S] [K_D]` but would instead expand to `[K_A] [K_B] [K_C] [K_D]`!

As long as this is explicitly documented somewhere.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 14.0.252-beta](https://keyman.com/downloads/releases/beta/14.0.252)

File: windows/src/global/vc/xstring.cpp [hunk 5]
```diff
@@ -47,7 +47,7 @@ PWSTR incxstr(PWSTR p)
 		case CODE_INDEX:		return p+2;
 		case CODE_USE:			return p+1;
 		case CODE_DEADKEY:		return p+1;
-		case CODE_EXTENDED:		p += 2; while(*p != UC_SENTINEL_EXTENDEDEND) p++; return p+1;
+		case CODE_EXTENDED:		p += 2; while(*p && *p != UC_SENTINEL_EXTENDEDEND) p++; return p+1;
```
Comment (mcdurdin) [inline windows/src/global/vc/xstring.cpp:50]: This fixes a buffer overrun on malformed xstrings. I need to make the same change in Keyman Engine and verify Keyman Core also.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.19-alpha](https://keyman.com/downloads/releases/alpha/15.0.19)

**Pull request merged**